src/pages/studyView/StudyViewUtils.tsx

```diff
@@ -3022,6 +2971,87 @@ export function updateSavedUserPreferenceChartIds(
     return chartSettings;
 }
 
+export async function getAllClinicalDataByStudyViewFilter(
+    studyViewFilter: StudyViewFilter
+): Promise<{ [sampleId: string]: { [attributeId: string]: string } }> {
+    const localClinicalDataCollection: ClinicalDataCollection = {
+        sampleClinicalData: [],
+        patientClinicalData: [],
+    };
+    let remoteClinicalDataCollection: ClinicalDataCollection = {
+        sampleClinicalData: [],
+        patientClinicalData: [],
+    };
+
+    const maxPageSize = 500000;
+    let pageNumber = 0;
+
+    do {
+        const remoteClinicalDataCollection = await internalClient.fetchClinicalDataClinicalTableUsingPOST(
+            {
+                studyViewFilter,
+                pageSize: maxPageSize,
+                pageNumber: pageNumber,
+                searchTerm: undefined,
+                sortBy: undefined,
+                direction: 'ASC',
+            }
+        );
+        localClinicalDataCollection.sampleClinicalData = localClinicalDataCollection.sampleClinicalData.concat(
+            remoteClinicalDataCollection.sampleClinicalData
+        );
+        localClinicalDataCollection.patientClinicalData = localClinicalDataCollection.patientClinicalData.concat(
+            remoteClinicalDataCollection.patientClinicalData
+        );
+        pageNumber++;
+    } while (remoteClinicalDataCollection.sampleClinicalData.length > 0);
+
+    return mergeClinicalDataCollection(localClinicalDataCollection);
+}
+
+export function mergeClinicalDataCollection(
+    clinicalDataCollection: ClinicalDataCollection
+): { [sampleId: string]: { [attributeId: string]: string } } {
+    const patientKeyedSampleData = _.groupBy(
+        clinicalDataCollection.sampleClinicalData,
+        d => d.uniquePatientKey
+    );
+    let patientKeyedSampleKeyedData = _.mapValues(
+        patientKeyedSampleData,
+        sampleData => _.groupBy(sampleData, d => d.uniqueSampleKey)
+    );
+    const patientKeyedPatientData = _.groupBy(
+        clinicalDataCollection.patientClinicalData,
+        d => d.uniquePatientKey
+    );
+    // Add patient level clinical data to sample clinical data.
+    patientKeyedSampleKeyedData = _.mapValues(
+        patientKeyedSampleKeyedData,
+        (sampleKeyedData, patientId) =>
+            _.mapValues(sampleKeyedData, (attrs, sampleId) =>
+                attrs.concat(patientKeyedPatientData[patientId] || [])
```

Comment:
it seems that this would duplicate the patient level items for each sample?

Reply:
No, this structure merely keys the sample-level clinical data by patient id like so:

```
{
  s1key: [{p1key: 'p1key'},]
}
```

becomes:

```
{
  p1key: {
     s1key: [{p1key: 'p1key'},]
  }
}
```
In essence, it collects the sample data per patient. This is done to later add the same patient data to each sample datum.